### ISSUE TITLE
Discussion about Plex limitations and possible solutions

### ISSUE THREAD

**Issue by ndbroadbent** [nicko88/HTFanControl#21]
Hello! I use Plex Media Server, and I run the Plex app on a Roku in my living room. I like to watch 4K movies with a high bitrate, so I don't think I can set up Kodi on a Raspberry Pi, and I don't want to buy another PC or NUC.

I'm curious about the limitations with Plex, how it only reports the current time to the nearest second. I want to look into this further and see if I can increase the accuracy. It's not important for wind or smoke effects, but I would love to get my strobe lights flashing at the right millisecond (explosions, gunshots, etc.)

I just found your Reddit post about this issue: https://www.reddit.com/r/PleX/comments/h17jea/plex_api_request_help_player_timestamp_and/

I'm curious to know if anything has changed, and if it's now possible to accurate timestamp information.

----

This seems to say that the time is returned in milliseconds: https://github.com/plexinc/plex-media-player/wiki/Remote-control-API#timeline-request

> time is the current playback position, in milliseconds.

**Comment by nicko88** [comment]
Yeah it's returned in milliseconds, but I can poll the player as fast as I want, and I don't get a new value any less than 1 second apart from the previous.  It just keeps returning the same time for 1+ seconds and then returns a new time for another second+.

Also my Plex player polling code might just not be very good because it often doesn't work for some of the people who have tried to use it.  Sometimes they just cant get it to connect to the plex player or get it to return data etc.  It always works for me, so I feel somewhat helpless to suggest what else they can try.  I have tried it on plex clients ranging from apple devices, roku, shield, and pc and it always at least connects and returns times for me.

The timing resolution of a Plex client seems to be random though.

Some clients are 1 second (HTPC), some are like 1.27 seconds (iPhone, iPad), some even 10 seconds (Nvidia Shield).  Although I think i had also seen 1 second on the nvidia shield as well.

The Plex Media Server timing resolution for connected clients seems to be 5 seconds.

You might think 1 second is OK for wind, but in my experience it's not good enough.  Me and my close friend (who I really developed the app for initially) have done some testing and we are really aiming for a seamless effect that is not gimmicky.  In our testing it seems like more than about 250ms is noticeable for wind, and when it's even just a little off like 500ms+, the effect becomes noticeable and not so "transparent" to the experience.

I am not sure how better to describe this, but if the air hits you at just the right moment, it just blends so seamlessly with the viewing experience, and if its off a little, it ends up being more of a distraction instead, pulling you out of the movie experience.

So in our opinion Plex is just not a viable option.  I have no idea how higher resolution timing can be had from it.

IMO you are better off using audio sync timing I think than plex.  With audio sync timing it's usually average accuracy around +/-100ms with worst cases occasionally around +/-250ms.

Nobody has extensively tested audio sync mode in a real theater setup so there could still be limitations to it's overall usability (background noise, too low volume, etc).  But it seems to work pretty well in my own testing with it.

**Comment by ndbroadbent** [comment]
Thanks for the detailed reply! I see what you mean about the wind effect timing, that makes a lot of sense. Unfortunately I can't use the audio sync mode either, since my server is in a closet on a different floor. But maybe I could set up a wireless microphone on an ESP32 board, and send the audio over WiFi.

I checked the timestamps for the Plex player on my iPhone, and it seems like it's sending timestamps with millisecond precision, e.g.:

* Current time: 00:04:16.817
* Current time: 00:04:18.832
* Current time: 00:00:03.125

And I noticed that Plex on my Roku was sending timestamps ending in `.500`, so it looks like it has 1 second resolution.

But it's all over the place and unpredictable - sometimes it hits the events right on time, and then other times there's a delay, or it sends the event too fast. But the spacing between the events is always accurate, so they are all being shifted up to one second.

I'm not going to use these effects every day since it's more of a "special occasion" kind of thing where I set it all up in our living room. So I might just set up Kodi on my Macbook and use that to play the movie. I will try that out now

**Comment by nicko88** [comment]
It's been awhile since I checked the times coming back from my iPhone.  They do update the app all the time so it's also possible it's changed.

**Comment by ndbroadbent** [comment]
I just got Kodi working and it's much better! I can even use the Kodi Plex plugin to access all my movies on the Plex server, but the timestamp is much more accurate when playing through Kodi. So I will use this for now.

I also bought some of these solid-state relay boards on AliExpress: https://www.aliexpress.com/item/33005213016.html
Hopefully these will allow me to silently control my fans and other devices.

Now I just need to find some fans! I think I'll go for 3 of the Honeywell HT-900 Fans. (The [Google Doc guide](https://docs.google.com/document/d/1Hdbm4cR5fi7vUoz1-eEbLAkaXpWckg9p/edit#heading=h.35nkun2) is really good!)

------

EDIT: That fan seems a bit small, and hard to find in New Zealand. It's actually pretty hard to find a fan! Some of them are really expensive and I'm not sure how much I want to spend. I might post on the AVR forum

**Comment by nicko88** [comment]
Excellent!

Solid state relays are a great idea too.  Regrettably the ones that a few people are using are fairly loud.  However they have the benefit of coming with an esp8266 onboard and with tasmota preinstalled.  And there's a step-by-step to get it fully configured for interlock mode and such.  So for some people it's much easier.

Also I didn't know much about the tasmota ecosystem and esp chips and such and that was all recommended by a user from the UK as well as the idea to replace the physical speed switches on an AC fan.

The very first couple users used IR controlled fans that just luckily happened to have discrete IR codes for each speed which was crucial.  However, that fan model was then discontinued only a few months after this project was created.  Also I don;t think it was ever available outside the USA either.

The google doc guide was written by another one of my home theater friends.  He keep detailed guides of everything he does related to home theater and it really is great.

Glad you have Kodi working.  There are actually 2 different Plex plugins for Kodi, and at least in my testing I thought the time sync was just as accurate.  At least I am even more confident about PlexKodiConnect which is the unofficial Plex plugin for Kodi.  The official one (called Plex for Kodi) should be just as accurate too, but maybe there is some reason it's may not be?  As far as I see it's Kodi doing all the playback and so the times should be the same as native Kodi.  With the exception of if you need media transcoding with Plex for Kodi, it wont really work because the start time is from when the transcoder starts, which can be way off if you don't start exactly at the beginning of the file.

https://www.avsforum.com/threads/4d-theater-wind-effect-diy-home-theater-project.3152346/page-59#post-61909214

Yeah, the AVS forum is a good place to continue this.

**Comment by ndbroadbent** [comment]
Hello again! I just realized that the Plex app was working great within Kodi and everything was synced correctly. I realized that the main difference was that I was connecting directly to the player (Kodi) instead of Plex Media Server.

So then I was thinking that I could probably do the same thing with my Roku, and get the timestamp information directly from the Roku IP while it’s running the Plex app. They have an API for this (which I had already set up in Home Assistant through the Roku integration): https://developer.roku.com/en-gb/docs/developer-program/debugging/external-control-api.md

Query/media-player example

The following command retrieves media player information.

```
$ curl "http://$ROKU_DEV_TARGET:8060/query/media-player"
```

The response includes the following fields:

```
<player error="false" state="play">
    <plugin bandwidth="44692475 bps" id="dev" name="MultiLive"/>
    <format audio="aac" captions="none" container="mp4" drm="none" video="mpeg4_15" video_res="1280x546"/>
    <buffering current="1000" max="1000" target="0"/>
    <new_stream speed="128000 bps"/>
    <position>6916 ms</position>
    <duration>887999 ms</duration>
    <is_live>false</is_live>
    <runtime>887999 ms</runtime>
    <stream_segment bitrate="0" media_sequence="1" segment_type="mux" time="0"/>
</player>
```

It looks like the position attribute should be quite accurate. I might look into this further and see if I can get it to work, but still having some problems compiling the app.

**Comment by nicko88** [comment]
This is a good find.  I never really looked closely into the Roku API to see what was available.

For a full solution you would need to pull the movie name from somewhere like the Plex API as well so that the loading of the proper movie happens.

Assuming the Roku API data turns out to indeed be truly "accurate" timecodes, a "hybrid" sync source method could be added to HTFanControl to pull the time from Roku and the movie name from Plex (if the movie name is not published anywhere in the Roku API).

I tried to help fix your build issues in the other issue in hopes you can start playing around with this.

**Comment by nicko88** [comment]
I took a stab at this, here are the code changes:

You should just be able to replace these files and build.  I didn't do much testing because I am not at home and don't have real access to a setup to properly test this on.

[HTFanControl_rokuplex.zip](https://github.com/nicko88/HTFanControl/files/10328145/HTFanControl_rokuplex.zip)

I also made you a build here for linux:

[HTFanControl_Linux_0.3.3.1.zip](https://github.com/nicko88/HTFanControl/files/10328148/HTFanControl_Linux_0.3.3.1.zip)

I do plan to work on this more tonight and can test it better as well as vet the timing accuracy. I can put out a beta build for you as well.

**Comment by ndbroadbent** [comment]
Oh thanks a lot! I was also able to get it to build on linux by removing the FanTrayIcon files. I attempted to also set up a "RokuPlex" player but your code is much better! The stopwatch code you added is very interesting, I didn't realize that was needed. I will test this out!

**Comment by nicko88** [comment]
Honestly I’m not totally sure it’s needed. 

But I just noticed when querying the Roku API that it takes between about 50-100ms to return a result which seems kinda slow to me, at least compared to querying MPC and Kodi. 

Though there’s no way to know what portion of or if any of that time is before the Roku gathers its player time and returns it or if it’s after. For now I just split it in half as I have a hunch it may be slightly more accurate. 

Tonight I am going to do more testing with some test videos to try to determine if any sort of adjustment is needed or helpful to the Roku API call.

**Comment by nicko88** [comment]
I tested the Roku timing, and it seems a little slow and a little inconsistent.  Currently it seems pretty good though if I add ~250ms to the queried time plus the full query stopwatch duration.

`VideoTime = Convert.ToInt64(Regex.Replace(strPosition, "[^0-9.]", "")) + 250 + sw.ElapsedMilliseconds;`

This is how it will be in the beta I release soon.

I test this with the global offset calibration file from here:  
https://nicko88.com/misc/HTFan/OffsetCalibration.zip

And I set all the fan offsets in HTFanControl to 0.  Then I make sure to have the program console window open in my same "eyesight" as TV playing the test video on the Roku through Plex app.  Then I can "see" how close the console logs the command compared to when the test video changes the TV to white or black.
